### PR TITLE
Dp/kh/api key

### PR DESCRIPTION
**Pull request by Dpalazzari**
In this PR: 

- Adds Api Key functionality
- Adds Api permissions
- Adjusts test to compensate for new api key functionality and permissions.
- All tests pass

### CODE REVIEW

**Comment by Dpalazzari** [comment]
@case-eee  @Robbie-Smith @drod1000 @jdconrad89 Ready for some review mateys.

File: app/controllers/api/v1/base_controller.rb [hunk 1]
```diff
@@ -1,4 +1,21 @@
...
+      if params[:controller] == "api/v1/comments"
+        redirect_to new_session_path
+      else
+        render json: {
+             error: "Invalid API key",
+             status: 400
+             }, status: 400
+      end
+    end
+  end
```
Comment (Robbie-Smith) [inline app/controllers/api/v1/base_controller.rb]: @kheppenstall @Dpalazzari What do you think about extracting lines 7 - 14 out into their own method, and then just calling that method?

Comment (Dpalazzari) [inline app/controllers/api/v1/base_controller.rb]: @Robbie-Smith Fixing it as I type this

File: app/services/api_permission.rb [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    return true if controller == 'api/v1/users/comments' && action.in?(%w(show))
+  end
+
+  def permitted_for_registered?
+    return true if controller == 'api/v1/comments' && action.in?(%w(index show update create destroy))
+    return true if controller == 'api/v1/projects/comments' && action.in?(%w(show))
+    return true if controller == 'api/v1/users/comments' && action.in?(%w(show))
+  end
+
+  def permitted_for_guest?
```
Comment (Robbie-Smith) [inline app/services/api_permission.rb]: @Dpalazzari @kheppenstall What do you both think about trimming down the logic in here? It seems a bit redundant to be me. The way it is written now, everyone has the same permissions except a guest. Couldn't we just make a before_action requiring someone to log in?

Comment (Dpalazzari) [inline app/services/api_permission.rb]: before_action on what @Robbie-Smith ?

File: spec/rails_helper.rb [hunk 3]
```diff
@@ -9,11 +9,11 @@
...
 require 'rack_session_access/capybara'
-require 'simplecov'
-SimpleCov.start
+# require 'simplecov'
+# SimpleCov.start
 
-require 'codecov'
-SimpleCov.formatter = SimpleCov::Formatter::Codecov
+# require 'codecov'
+# SimpleCov.formatter = SimpleCov::Formatter::Codecov
```
Comment (Robbie-Smith) [inline spec/rails_helper.rb]: Any reason lines 12-16 are commented out?

Comment (Dpalazzari) [inline spec/rails_helper.rb]: @Robbie-Smith It was causing some problems when rspecing so we commented them out. Will revert now. Good catch. 👍

File: app/services/api_permission.rb (hunk 2 above)
Comment (Robbie-Smith) [inline app/services/api_permission.rb]: @Dpalazzari on the api/v1/basecontroller

File: spec/requests/api/v1/comments/comments_update_spec.rb [hunk 4]
```diff
@@ -6,11 +6,10 @@
     project = create(:project)
     comment = create(:comment, project: project, user: user)
 
-    expect_any_instance_of(Api::V1::BaseController).to receive(:current_user).and_return(user)
-    expect_any_instance_of(Api::V1::BaseController).to receive(:current_user).and_return(user)
+    stub_login(user)
```
Comment (Robbie-Smith) [inline spec/requests/api/v1/comments/comments_update_spec.rb]: @Dpalazzari @kheppenstall I like how you guys cleaned up this test.

File: app/services/api_permission.rb (hunk 2 above)
Comment (kheppenstall) [inline app/services/api_permission.rb]: @Robbie-Smith I'm unsure of how we would pull out the logic of the roles, login, and api key into the controller without having tons of methods in the base controller. To me it seems like a PORO is a better way to handle all the different cases - users having different roles, users being logged in or not, and the api key coming through in the params.

Comment (Robbie-Smith) [inline app/services/api_permission.rb]: @kheppenstall I agree and I like how you and @Dpalazzari refactored.

**Comment by Robbie-Smith** [comment]
@Dpalazzari @kheppenstall merged

**Pull request merged**